mpeg2/src/ts.rs

```diff
@@ -36,21 +39,66 @@ impl AdaptationField {
             } else {
                 None
             };
+            let mut n = if af.program_clock_reference_27mhz.is_some() { 8 } else { 2 };
+            if buf[1] & 0x02 != 0 {
+                let transport_private_date_length = buf[n];
+                af.private_data_types = buf[n..(n + transport_private_date_length as usize)].to_vec();
+                n += transport_private_date_length as usize + 1
+            }
+            if buf[1] & 1 == 1 {
+                if buf[n] == 0 {
+                    return Err(DecodeError::new("invalid adaptation_field_extension_length"));
+                }
+                let temi_descriptors_len = (buf[n] - 1) as usize;
+                n += 2;
+                af.temi_timeline_descriptors = vec![];
+                let end = n + temi_descriptors_len;
+                while n < end {
+                    if buf[n] == AF_DESCR_TAG_TIMELINE {
+                        let descr = TEMITimelineDescriptor::decode(&buf[n..end])?;
+                        n += descr.encode_len();
+                        af.temi_timeline_descriptors.push(descr);
+                    } else {
+                        // Other types of TEMI descriptors are ignored
+                        n += buf[n + 1] as usize;
+                    }
+                }
+                if n != end {
+                    return Err(DecodeError::new("not enough bytes for temi timeline descriptors"));
+                }
+            }
         }
         Ok(af)
     }
 
     pub fn encoded_len(&self) -> usize {
-        1 + if self.discontinuity_indicator.is_some() || self.random_access_indicator.is_some() || self.program_clock_reference_27mhz.is_some() {
-            1
-        } else {
-            0
-        } + if self.program_clock_reference_27mhz.is_some() { 6 } else { 0 }
+        let mut len = 1;
+        if self.discontinuity_indicator.is_some() || self.random_access_indicator.is_some() || self.program_clock_reference_27mhz.is_some() {
+            len += 1;
+        }
+        if self.program_clock_reference_27mhz.is_some() {
+            len += 6;
+        }
+        if !self.private_data_types.is_empty() {
+            len += 1 + self.private_data_types.len();
+        }
```

Comment:
Don't we also need to add 2 bytes for `adaptation_field_extension_length` and the flags after it?

```suggestion
        }
        if !self.temi_timeline_descriptors.is_empty() {
                len += 2;
        }
```